Use canonical constant time comparison

paragonie-scott opened the pull request:
https://cryptocoding.net/index.php/Coding_rules#Compare_secret_strings_in_constant_time

brianmhunt commented on constantTimeCompare.js:
```diff
@@ -30,18 +30,17 @@ module.exports = function constantTimeCompare(a, b) {
+        match = 0,
         i = Math.max(aLen, bLen, MAX_KEY_CHARS);
+    // If the lengths don't match, the string doesn't match
+    if (aLen !== bLen) {
```
You can't early out of a constant-time comparison.

paragonie-scott replied:
The current way looks like a naive way to "not leak the string length", which is something you can't do because your runtime will always leak that. Also, string lengths are public and not useful for attackers.

> Ha! We figured out that their SHA256 hash is exactly 32 bytes long! That buys us....... nothing.

brianmhunt commented on constantTimeCompare.js:
```diff
@@ -30,18 +30,17 @@ module.exports = function constantTimeCompare(a, b) {
-      // We repeat the comparison over the strings with % so that we do not compare
-      // a number to NaN, since that has different timing that comparing two numbers.
-      match &= a.charCodeAt( i % aLen ) === b.charCodeAt( i % bLen ) ? 1 : 0;
+      match |= a.charCodeAt( i ) ^ b.charCodeAt( i );
```
This has a number of problems.

1. the var `i` is over the longer strings' length.  Supposing `a` is shorter, then `a.charCodeAt` takes a different code-path for the difference in length between the strings.  This code-path is identifiable.

2. `"123".charCodeAt(4)` returns `NaN`, which comparison takes a different code path than comparing two ints (i.e. the comparison fails on strict type equality before it fails on value equality).  You should've noted this from the comments this patch deletes.

paragonie-scott replied:
> This has a number of problems.

Neither of those will never happen, because we exit early when the lengths differ.

brianmhunt replied:
— Got it.

As this is not a general-purpose constant-time comparison it may be appropriate to assume that the strings will be of equal length (i.e. it's only going to be comparing strings from the same hash function in the ordinary course) may be one we're prepared to make.

I'm not positive that we can assume equal string lengths, but if we can then I think this patch makes sense.

To make sure it is clear, I think the function name should change or a big comment should be added to indicate that it is only constant-time for strings of equal length and otherwise short-circuits.

brianmhunt replied:
... in the interest of assurances that the line lengths being compared are equal, I feel that when strings of different lengths are provided to the function the proper thing to do is to throw an error.

brianmhunt commented on constantTimeCompare.js:
```diff
@@ -30,18 +30,17 @@ module.exports = function constantTimeCompare(a, b) {
-        match = aLen === bLen ? 1 : 0,
+        match = 0,
         i = Math.max(aLen, bLen, MAX_KEY_CHARS);
+    // If the lengths don't match, the string doesn't match
```
This comment just repeats what the next line of code says, so I think it can safely be removed.

mastilver commented:
@brianmhunt Except the comment, does everything else looks good to you?

brianmhunt commented:
I think this'll do.

As a matter of interest here's the bitwise comparison from [SJCL](https://github.com/bitwiseshiftleft/sjcl/blob/23cd0dee310a339b5b40e32924f0e91cd27967c6/core/bitArray.js#L136):

```javascript
   * Compare two arrays for equality in a predictable amount of time.
   * @param {bitArray} a The first array.
   * @param {bitArray} b The second array.
   * @return {boolean} true if a == b; false otherwise.
   */
  equal: function (a, b) {
    if (sjcl.bitArray.bitLength(a) !== sjcl.bitArray.bitLength(b)) {
      return false;
    }
    var x = 0, i;
    for (i=0; i<a.length; i++) {
      x |= a[i]^b[i];
    }
    return (x === 0);
  }
```

As @paragonie-scott note's, what credential does now is a naïve string-length masking, and even if we leak the string length it's not generally considered useful.

mastilver commented:
@brianmhunt Why don't we use that library then?

mastilver commented:
@paragonie-scott Could you rebase with master so we are sure unit test are passing

@brianmhunt feel free to merge it when you are happy ;)

ericelliott commented:
Closing all stale issues and PRs. Please update, ensure all tests pass, and reopen if you really want this in.

paragonie-scott commented:
I don't "really want" *anything* in, this won't affect me. When I touch node.js, I just use libsodium.

If you want your constant-time comparison loop to be in line with what other cryptography libraries do, consider applying the patch I sent.

ericelliott commented:
Interesting.

Libsodium uses Argon2 as the default password hashing algorithm. Do you use it?

AFAIK, it's main advantage over PBKDF2 is that it's "memory hard". I'm not sure I see the huge advantage there, considering that attacks from single systems running lots of GPUs are far less common than botnet cracks (IoT botnets have ~1 million zombies -- all with their own memory -- to throw at the problem).

Other than that, PBKDF2 is far more commonly used, and has survived far more peer scrutiny (~2k scholarly research articles compared to ~500 for Argon2). Argon2 did win a contest -- but how many researchers participated in the reviews, and was PBKDF2 compared in that contest?

paragonie-scott commented:
> Libsodium uses Argon2 as the default password hashing algorithm. Do you use it?

Used, implemented, attempted cryptanalysis. I've done it all. I've also implemented and tried to break BLAKE2 (which Argon2 uses internally).

> I'm not sure I see the huge advantage there, considering that attacks from single systems running lots of GPUs are far less common than botnet cracks (IoT botnets have ~1 million zombies -- all with their own memory -- to throw at the problem).

Are you familiar with the scrypt paper? https://www.tarsnap.com/scrypt/scrypt.pdf

I don't have current hardware numbers on the efficacy of GPU clusters over CPU cracking PBKDF2 versus bcrypt, scrypt, or Argon2. However, [this StackOverflow answer](https://security.stackexchange.com/a/32845/43688) should provide some insight into *why* memory hardness offers a security advantage.

> Argon2 did win a contest -- but how many researchers participated in the reviews, and was PBKDF2 compared in that contest?

It didn't win a contest, it won a multi-year cryptography **competition**, open to the public, to assess memory-hard password hashing and/or key derivation functions in anticipation of advances against bcrypt and scrypt.

The distinction matters; this wasn't a trivial academic undertaking by some obscure committee. The password hashing competition was structured similarly to the AES and SHA3 competitions, and involved a lot of well-renowned cryptographers with special interest in both fast cryptographic hash functions and password hash functions (including Colin Percival, JP Aumasson, Peter Gutmann, and Alexander Peslyak).

Yescrypt, one of the finalists, is based on (and compatible with) scrypt, which uses PBKDF2 as an extensible length output function (although its security properties largely depend on Salsa20/8 as implemented in ROMix). In that sense, sure, PBKDF2 was "evaluated". I'm pretty sure the real measuring sticks were bcrypt and scrypt, however. PBKDF2 wasn't considered adequate for the job.

The advantage of PBKDF2 is that it's a simple construction (Secure hash function + HMAC + a little bit of glue gives you a standardized iterated hash function deemed suitable for password hashing and key derivation), whereas bcrypt is only suitable for password hashing, and scrypt isn't widely available.

ericelliott commented:
Thanks for your thoughtful reply. Very interesting! I come from the perspective of a developer building apps that we have to protect from real attackers, so I'm interested in which algorithms provide the best protections from the actual threats we see in the wild, rather than theoretical performance.

I am aware of the scrypt paper and memory hard functions, but I am not an expert by any means. I'm asking these questions because I'm trying to understand if we should be recommending something like Argon2 instead of PBKDF2.

When it comes to security, I like to work with as much data as I can, and I'm simply pretty ignorant of the details of how much research went into Argon2, and whether or not it was tested against common real world attack vectors like Botnets (and of course, GPUs), and if so, how it stands up compared to PBKDF2. If you know any references to comparative research like that, I'd be very grateful.

> I don't have current hardware numbers on the efficacy of GPU clusters over CPU cracking PBKDF2 versus bcrypt, scrypt, or Argon2. However, this StackOverflow answer should provide some insight into why memory hardness offers a security advantage.

I understand why Argon2 offers an advantage on a single machine with GPUs. I don't understand why it would offer an advantage over the biggest password cracking threat: Botnets.

[Botnets are commonly used in password cracking](https://www.cso.com.au/article/630001/how-hackers-crack-passwords-why-can-t-stop-them/) in the wild today, as I mentioned with up to ~1 million zombie devices or more in a single coordinated effort. Most of those devices are cheap with little or no GPU power. On those botnets, parallelism is network bound, and botnets don't need GPUs to be effective. IoT botnets are generally memory constrained, but zombie windows machines often have gigs of memory to spare. Parallelism is spread across many machines with their own memory.

Does Argon2 offer any advantages against those common attacks?

paragonie-scott commented:
> Does Argon2 offer any advantages against those common attacks?

It can, if you set the memory factor to be larger than what most IoT devices have available, yes.

ericelliott commented:
> It can, if you set the memory factor to be larger than what most IoT devices have available, yes.

This sounds like it could be effective against a subclass of IoT botnets, but some enterprise fog/[IIoT devices](https://www.iotforall.com/intelligent-iot-fog-computing-trends/) have RAM comparable to common hosted virtual machines where a password authentication server might reside (4-16 GB).

Wouldn't it be prohibitively expensive to set requirements capable of impeding botnets made up of things like compromised IIoT switches?

ericelliott commented:
Thanks for your patience. I appreciate your input.

I've searched for more detail on the [password hashing competition that Argon2 won](https://password-hashing.net/), and found very little useful information about the competition methodology, how many researchers were involved, links to resulting research papers, details on experiments performed, etc...

The public mailing list link is broken.

Do you have any links where I can learn more about Argon2? As I mentioned, I'm very interested in making a data-informed decision, and there is a wealth of detail available on PBKDF2 research, as well as my own experience using it in production with hundreds of millions of users.

ericelliott commented:
Re: your comment:

> It would be really helpful if you picked one threat model at a time.

That's not how botnets work. Most million+ zombie botnets have modular attack vectors and infect a variety of device classes. The limited-memory devices tend to be employed for tasks like DDoS attacks, while the beefier devices (IIoT switches, infected servers, infected Windows machines) are generally the ones used for password cracking and cryptojacking.

Sc00bz commented:
This commit looks good. The only thing I'd change because aLen is the same as bLen and no one limits max length of a general compare function:
```
- i = Math.max(aLen, bLen, MAX_KEY_CHARS);
+ i = aLen;
```

Also the original code is not constant time due to branch prediction.
```
a.charCodeAt( i % aLen ) === b.charCodeAt( i % bLen ) ? 1 : 0
```

> PBKDF2 is far more commonly used, and has survived far more peer scrutiny

That only works if you ignore the research that points out its flaws. PBKDF2 has a footgun and this actually went off in this project. See #83 . Also this isn't a dig against you Microsoft and 1Password fell for this footgun... and kinda Google. I like bringing up Google Sync when talking about the PBKDF2 footgun because it's just so bad. It's like they took the PBKDF2 footgun and added a chainsaw bayonet. Shot their foot and to fix the gun shot wound they amputated their foot with said chainsaw. Then called closed all bug reports as "won't fix". Note before I reported it these bugs where reported 2 or 3 times each. Anyway don't be like Google.

For authentication nothing beats pre-hashed (ASCII encoded) bcrypt: "bcrypt(hex(sha256(pw)), salt)". Pre-hashed (ASCII encoded) fixes all bugs in all previous and current versions of bcrypt. Also avoids the null character footgun. Hash and encoding don't matter. As long as it's not like MySQL323, CRC\*, or ASCII binary (literal "0" and "1"). When looking at time to hash vs cracking speed bcrypt is 85x better than PBKDF2-SHA1 and PBKDF2-SHA256 and 40x better than PBKDF2-SHA512. Bcrypt is better than Argon2 and scrypt when less than around 0.1 second, but bcrypt is worse when more than around 1 second.

Anyway min settings for "awesome": PBKDF2-SHA1 500k, PBKDF2-SHA256 190k, PBKDF2-SHA512 70k, and bcrypt 7. These get you <10kH/s/GPU. That said bcrypt 8 is the minimum I'd recommend but only because it's still fast. Note most people do 10 or higher. In 2000, PBKDF2's minimum was 1k for SHA1 and you should "double every 2 years"\* and then GPUs 10x-ed its speed. So it's 1k\*2^9\*10 or about 5000k--oh, so 10x those numbers.

\* Moore's law lately has slowed, but for specific tasks GPUs made up the slack and then some but have also recently slowed a little. Oh crazy it even made it into the [Wikipedia page: "Intel stated in 2015"](https://en.wikipedia.org/wiki/Moore%27s_law). Wait that's where I say that happened: [skylake](https://en.wikipedia.org/wiki/Template:Intel_processor_roadmap). Obviously they said it first and they have better... intel.

tjconcept commented:
FYI Node.js now has a [`crypto.timingSafeEqual(a, b)`](https://nodejs.org/api/crypto.html#crypto_crypto_timingsafeequal_a_b) function.

ericelliott commented:
I'm open to a PR for `crypto.timingSafeEqual`.
